library/spdm_common_lib/libspdm_com_opaque_data.c

```diff
@@ -401,64 +440,111 @@ libspdm_process_opaque_data_version_selection_data(libspdm_context_t *spdm_conte
         .spdm_version_count == 0) {
         return LIBSPDM_STATUS_SUCCESS;
     }
 
-    if (data_in_size !=
-        libspdm_get_opaque_data_version_selection_data_size(spdm_context)) {
-        return LIBSPDM_STATUS_INVALID_MSG_SIZE;
-    }
     if (libspdm_get_connection_version (spdm_context) >= SPDM_MESSAGE_VERSION_12) {
         spdm_general_opaque_data_table_header = data_in;
-        if (spdm_general_opaque_data_table_header->total_elements != 1) {
+        if (spdm_general_opaque_data_table_header->total_elements < 1) {
             return LIBSPDM_STATUS_INVALID_MSG_FIELD;
         }
         opaque_element_table_header =
             (void *)(spdm_general_opaque_data_table_header + 1);
+
+        element_num = spdm_general_opaque_data_table_header->total_elements;
     } else {
         general_opaque_data_table_header = data_in;
         if ((general_opaque_data_table_header->spec_id !=
              SECURED_MESSAGE_OPAQUE_DATA_SPEC_ID) ||
             (general_opaque_data_table_header->opaque_version !=
              SECURED_MESSAGE_OPAQUE_VERSION) ||
-            (general_opaque_data_table_header->total_elements != 1)) {
+            (general_opaque_data_table_header->total_elements < 1)) {
             return LIBSPDM_STATUS_INVALID_MSG_FIELD;
         }
         opaque_element_table_header =
             (void *)(general_opaque_data_table_header + 1);
+
+        element_num = general_opaque_data_table_header->total_elements;
     }
-    if ((opaque_element_table_header->id != SPDM_REGISTRY_ID_DMTF) ||
-        (opaque_element_table_header->vendor_len != 0) ||
-        (opaque_element_table_header->opaque_element_data_len !=
-         sizeof(secured_message_opaque_element_version_selection_t))) {
-        return LIBSPDM_STATUS_INVALID_MSG_FIELD;
-    }
-    opaque_element_version_section = (void *)(opaque_element_table_header + 1);
-    if ((opaque_element_version_section->sm_data_version !=
-         SECURED_MESSAGE_OPAQUE_ELEMENT_SMDATA_DATA_VERSION) ||
-        (opaque_element_version_section->sm_data_id !=
-         SECURED_MESSAGE_OPAQUE_ELEMENT_SMDATA_ID_VERSION_SELECTION)) {
-        return LIBSPDM_STATUS_INVALID_MSG_FIELD;
-    }
 
-    for (secured_message_version_index = 0;
-         secured_message_version_index <
-         spdm_context->local_context.secured_message_version.spdm_version_count;
-         secured_message_version_index++) {
-
-        if (libspdm_get_version_from_version_number(opaque_element_version_section->
-                                                    selected_version)
-            ==
-            libspdm_get_version_from_version_number(spdm_context->local_context.
-                                                    secured_message_version
-                                                    .spdm_version[secured_message_version_index]))
-        {
-            return LIBSPDM_STATUS_SUCCESS;
+    for (element_index = 0; element_index < element_num;
+         element_index++) {
+
```

Comment:
Since most checks are duplicated, would you please consider create a function to get the opaque_element via sm_data_id?

Reply:
I create function: `libspdm_process_opaque_data` , which can be called by `libspdm_process_opaque_data_version_selection_data` and `libspdm_process_opaque_data_supported_version_data`.
There is a parameter: `process_supported_version` to distinguish.

And I have moved the unit test on libspdm_common unit test.

Reply:
This is not a good idea. What if there are more opaque data need to handle?
A BOOL cannot handle more than 2 state.

Why not take my feedback to create a common function to return element based upon ID ?

Reply:
Because the many check is over before getting sm_data_id, I think the `libspdm_process_opaque_data` can be named `libspdm_process_opaque_data_negotiate_vesion`.

Reply:
I think we need split the ID specific check from common element check.

The common element check goes to common function.
The ID specific check still stays in original function.